File: src/test/java/com/github/qky666/selenidepom/test/java/mtp/pom/containerpom/MainFramePage.java
package com.github.qky666.selenidepom.test.java.mtp.pom.containerpom;

import com.codeborne.selenide.SelenideElement;
import com.github.qky666.selenidepom.config.SPConfig;
import com.github.qky666.selenidepom.pom.Loadable;
import com.github.qky666.selenidepom.pom.Required;
import org.openqa.selenium.support.FindBy;

import static com.codeborne.selenide.ClickOptions.withOffset;
import static com.github.qky666.selenidepom.pom.LoadableKt.hasLoadedRequired;
import static com.github.qky666.selenidepom.pom.LoadableKt.shouldLoadRequired;

public class MainFramePage implements Loadable {

    // Fields
    @FindBy(css = "a.img-menu") @Required public SelenideElement home;
    @FindBy(css = "nav.menu-pc") @Required(model = "desktop") public MainMenuWidget mainMenu;
    @FindBy(css = "button.custom-menu-btn-flotante") @Required(model = "mobile") public SelenideElement mobileMenuButton;
    @FindBy(css = "div#menu-movil ul.uk-nav") public MobileMenuWidget mobileMenu;
    @FindBy(css = "div#cookie-law-info-bar") public CookiesBannerWidget cookiesBanner;

    // Methods
    public void acceptCookies() {
        // This is what it should be, but for some reason, cookies message is not displayed when page loads
        // shouldLoadRequired(this).getCookiesBanner().acceptCookies();

        // Workaround
        if (SPConfig.INSTANCE.getModel().equals("mobile")) {
            acceptCookiesMobile();
        } else {
            acceptCookiesDesktop();
        }
    }

    private void acceptCookiesDesktop() {
        do {
            mainMenu.openSearch.click();
            mainMenu.langEs.click(withOffset(0, -50));
        } while (!hasLoadedRequired(cookiesBanner));
        cookiesBanner.acceptCookies();
        shouldLoadRequired(this);
    }

    private void acceptCookiesMobile() {
        shouldLoadRequired(this);
        do {
            mobileMenuButton.click();
        } while (!hasLoadedRequired(cookiesBanner));
        cookiesBanner.acceptCookies();
    }
}

File: src/test/kotlin/com/github/qky666/selenidepom/test/kotlin/mtp/pom/MainFramePage.kt
package com.github.qky666.selenidepom.test.kotlin.mtp.pom

import com.codeborne.selenide.ClickOptions
import com.codeborne.selenide.Selenide.element
import com.github.qky666.selenidepom.pom.Required
import com.github.qky666.selenidepom.config.SPConfig
import com.github.qky666.selenidepom.pom.Page
import com.github.qky666.selenidepom.pom.shouldLoadRequired
import com.github.qky666.selenidepom.test.kotlin.mtp.pom.menu.desktop.MainMenuWidget
import com.github.qky666.selenidepom.test.kotlin.mtp.pom.menu.mobile.MobileMenuWidget

open class MainFramePage : Page() {
    @Required("desktop") val mainMenu = MainMenuWidget(element("div.custom-menu"))
    @Required("mobile") val mobileMenuButton = element("button.custom-menu-btn-flotante")
    val mobileMenu = MobileMenuWidget(element("div#menu-movil ul.uk-nav"))
    val cookiesBanner = CookiesBannerWidget(element("div#cookie-law-info-bar"))

    fun acceptCookies() {
        // This is what it should be, but for some reason, cookies message is not displayed when page loads
        // shouldLoadRequired().cookiesBanner.acceptCookies()

        // Workaround
        if (SPConfig.pomVersion == "mobile") {
            acceptCookiesMobile()
        } else {
            acceptCookiesDesktop()
        }
    }

    private fun acceptCookiesDesktop() {
        mainMenu.searchOpen.click()
        mainMenu.langEs.click(ClickOptions.withOffset(0, -50))
        cookiesBanner.acceptCookies()
        shouldLoadRequired()
    }

    private fun acceptCookiesMobile() {
        shouldLoadRequired().mobileMenuButton.click()
        cookiesBanner.acceptCookies()
    }
}

val mainFramePage = MainFramePage()
